### PR TITLE
le_tc/task: Add corner cases for task TC's

### PR DESCRIPTION
**Pull request by thapav**
This patch adds corner cases for following API's
task_create()
task_reparent()
task_restart()

Signed-off-by: Vidisha Thapa <thapa.v@samsung.com>

### CODE REVIEW

**Comment by seinfra** [comment]
Code Rule Check Result:
apps/examples/testcase/le_tc/kernel/tc_task.c:294: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
apps/examples/testcase/le_tc/kernel/tc_task.c:300: ERROR: [BRC_M_SMT] else should follow close brace '}'

File: apps/examples/testcase/le_tc/kernel/tc_task.c [hunk 1]
```diff
@@ -31,11 +31,13 @@
 #include <sys/types.h>
 #include <unistd.h>
 #include "tc_internal.h"
+#include <../kernel/sched/sched.h>
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_task.c]: Don't call kernel internal header.

File: apps/examples/testcase/le_tc/kernel/tc_task.c [hunk 2]
```diff
@@ -77,6 +97,24 @@ static int delete_task(int argc, char *argv[])
...
 }
 
+/**
+* @fn                   :reparent_task
+* @brief                :utility function for tc_task_task_reparent
+* @return               :int
+*/
+static int reparent_task(int argc, char *argv[])
+{
+
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_task.c]: remove empty line

File: apps/examples/testcase/le_tc/kernel/tc_task.c [hunk 3]
```diff
@@ -235,11 +273,38 @@ static int vfork_task(int argc, char *argv[])
...
+	/* Regular functionality check */
+
 	pid = task_create("tc_task_create", SCHED_PRIORITY_MAX - 1, 1024, create_task, (char * const *)task_param);
 	TC_ASSERT_GT("task_create", pid, 0);
 	TC_ASSERT_EQ("task_create", g_callback, OK);
+
+	/* New task creation failure check due to maximum task creation limit */
+
+	for(i = g_alive_taskcount; i <= CONFIG_MAX_TASKS; i++) {
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_task.c]: Same as previous. Don't call kernel internal variable.

File: apps/examples/testcase/le_tc/kernel/tc_task.c [hunk 4]
```diff
@@ -468,14 +586,15 @@ static void tc_task_vfork(void)
...
-	tc_task_exit();
 	tc_task_atexit();
+	tc_task_exit();
+	tc_task_getpid();
 	tc_task_on_exit();
 	tc_task_prctl();
-	tc_task_getpid();
+	tc_task_task_create();
+	tc_task_task_delete();
+	tc_task_task_reparent();
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/kernel/tc_task.c]: task_reparent is not a public api. I think we don't need to make not public api test case

Comment (thapav) [reply apps/examples/testcase/le_tc/kernel/tc_task.c]: ok, then for code coverage what should we do ?

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by thapav** [comment]
modified as per review comments

File: apps/examples/testcase/le_tc/kernel/tc_task.c (hunk 4 above)
Comment (jeongchanKim) [reply apps/examples/testcase/le_tc/kernel/tc_task.c]: only calling with public api, we can cover all internal apis.

**Comment by sunghan-chang** [comment]
@thapav As I and Jeongchan said, Don't call kernel internal APIs from apps.

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by thapav** [comment]
@sunghan-chang modified as per review comments, please check

**Comment by seinfra** [comment]
Code Rule Check OK.

File: apps/examples/testcase/le_tc/kernel/tc_task.c [hunk 5]
```diff
@@ -224,6 +239,46 @@ static int vfork_task(int argc, char *argv[])
...
+	int i;
+	pid_t pid = 0;
+	ppid = getpid();
+
+	/* intentionally creates more than CONFIG_MAX_TASKS */
+	for (i = 0; i < CONFIG_MAX_TASKS; i++) {
+		pid = task_create("tc_vfork_temp", SCHED_PRIORITY_MAX - 1, 1024, create_temp_task, (char * const *)NULL);
+	}
+
+	if (i == (CONFIG_MAX_TASKS)) {
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_task.c]: I don't understand a logic of checking a result. 
i.e.
We have 5 tasks and CONFIG_MAX_TASKS is 10. If this makes 4 tasks and gets ERROR as return. This TC will be passed.

How about check a number of task which is alive using sched_foreach?

File: apps/examples/testcase/le_tc/kernel/tc_task.c [hunk 6]
```diff
@@ -238,9 +293,24 @@ static void tc_task_task_create(void)
...
+	TC_ASSERT_EQ("task_create", errno, EINVAL);
+
+	/* Regular functionality check */
+
 	pid = task_create("tc_task_create", SCHED_PRIORITY_MAX - 1, 1024, create_task, (char * const *)task_param);
 	TC_ASSERT_GT("task_create", pid, 0);
 	TC_ASSERT_EQ("task_create", g_callback, OK);
+
+	/* New task creation failure check due to maximum task creation limit (>CONFIG_MAX_TASKS)*/
+	pid = task_create("tc_maxtask", SCHED_PRIORITY_MAX - 1, 1024, max_task, (char * const *)NULL);
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_task.c]: You make a task to test making maximun task. But that's all. There is no checking result logic (no TC_ASSERT macro).

File: apps/examples/testcase/le_tc/kernel/tc_task.c [hunk 7]
```diff
@@ -282,6 +354,18 @@ static void tc_task_task_restart(void)
 	int ret_chk;
 	unsigned int remain;
 	g_callback = 0;
+
+	/* Check for NULL pid parameter  */
+
+	pid = task_create("tc_exit", SCHED_PRIORITY_MAX - 1, 1024, restart_task, (char * const *)NULL);
+	TC_ASSERT_GT("task_create", pid, 0);
+
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_task.c]: 1. Let's check g_callback value to confirm that task is not restarted.
```
	pid = task_create("tc_exit", SCHED_PRIORITY_MAX - 1, 1024, restart_task, (char * const *)NULL);
	TC_ASSERT_GT("task_create", pid, 0);

	g_callback = 0;
	ret_chk = task_restart(0);
	TC_ASSERT_EQ("task_restart", ret_chk, ERROR);
	TC_ASSERT_EQ("task_restart", errno, ENOSYS);
	TC_ASSERT_EQ("task_restart", g_callback, 0);
```
2. If you added ```g_callback = 0; like above, you should remove line 356. It is unnecessary.
3. After test this, you should reinitialize g_callback to test next.

File: apps/examples/testcase/le_tc/kernel/tc_task.c [hunk 8]
```diff
@@ -183,6 +184,20 @@ static int vfork_temp_task(int argc, char *argv[])
...
+static int create_temp_task(int argc, char *argv[])
+{
+	int rc = 0;
+
+	waitpid(ppid, &rc, 0);
+
+	struct timespec st_timespec;
+	st_timespec.tv_sec = 0;
+	st_timespec.tv_nsec = 1;
+	nanosleep(&st_timespec, NULL);
```
[the remark this answers is not in the document]
Comment (thapav) [reply apps/examples/testcase/le_tc/kernel/tc_task.c]: I am able to meet max tasks number with this, so i suppose yes

File: apps/examples/testcase/le_tc/kernel/tc_task.c (hunk 5 above)
Comment (thapav) [reply apps/examples/testcase/le_tc/kernel/tc_task.c]: yes that can be done, but then it needs to be done under 
#ifndef CONFIG_BUILD_PROTECTED flag
This TC logic just focuses on achieving tasks greater than CONFIG_MAX_TASKS, so independent of current alive tasks, 
whenever this threshold is crossed, pid = ERROR 
after this task exits all tasks after ( i = CONFIG_MAX_TASKS ) return PID = ERROR &
all tasks created before ( i =CONFIG_MAX_TASKS) return valid PID's

Hence we dont need dependency over BUILD TYPE

**Comment by thapav** [comment]
@sunghan-chang modified as per review comments please check

**Comment by seinfra** [comment]
Code Rule Check OK.

File: apps/examples/testcase/le_tc/kernel/tc_task.c [hunk 9]
```diff
@@ -79,7 +80,7 @@ static int delete_task(int argc, char *argv[])
 }
 
 /**
-* @fn                   :restart_task
+k @fn                   :restart_task
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_task.c]: revert

File: apps/examples/testcase/le_tc/kernel/tc_task.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (sunghan-chang) [reply apps/examples/testcase/le_tc/kernel/tc_task.c]: I see. Due to waitpid, you can see max tasks. That means you don't need to call sleep function. Could you test it?

File: apps/examples/testcase/le_tc/kernel/tc_task.c (hunk 5 above)
Comment (sunghan-chang) [reply apps/examples/testcase/le_tc/kernel/tc_task.c]: 1. You didn't try to make tasks over CONFIG_MAX_TASKS. At end of TC, you didn't check how many tasks are alive. How do you say that is checking of max tasks?
2. Kishore made kernel test driver to run kernel side tc inside kernel. You should move kernel tc into kernel. Then you don't need to concern protected build.

File: apps/examples/testcase/le_tc/kernel/tc_task.c [hunk 10]
```diff
@@ -224,6 +239,40 @@ static int vfork_task(int argc, char *argv[])
...
+		pid = task_create("tc_vfork_temp", SCHED_PRIORITY_MAX - 1, 1024, create_temp_task, (char * const *)NULL);
+	}
+
+	if (i == (CONFIG_MAX_TASKS)) {
+		if (pid == ERROR) {
+			return OK;
+		} else {
+			return -1;
+		}
+	} else {
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_task.c]: When can we meet this else case? I think this is unreachable.

File: apps/examples/testcase/le_tc/kernel/tc_task.c [hunk 11]
```diff
@@ -238,9 +287,25 @@ static void tc_task_task_create(void)
...
+
 	pid = task_create("tc_task_create", SCHED_PRIORITY_MAX - 1, 1024, create_task, (char * const *)task_param);
 	TC_ASSERT_GT("task_create", pid, 0);
 	TC_ASSERT_EQ("task_create", g_callback, OK);
+
+	/* New task creation failure check due to maximum task creation limit (>CONFIG_MAX_TASKS)*/
+	pid = task_create("tc_maxtask", SCHED_PRIORITY_MAX - 1, 1024, max_task, (char * const *)NULL);
+	TC_ASSERT_GT("task_create", pid, OK);
+
+	sleep(SEC_3);
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_task.c]: As I left before, you didn't check any result of making max tasks. Please add checking logic of that with TC_ASSERT macro.

File: apps/examples/testcase/le_tc/kernel/tc_task.c [hunk 12]
```diff
@@ -282,6 +349,16 @@ static void tc_task_task_restart(void)
 	int ret_chk;
 	unsigned int remain;
 	g_callback = 0;
+
+	/* Check for NULL pid parameter  */
+
+	ret_chk = task_restart(0);
+	TC_ASSERT_EQ("task_restart", ret_chk, ERROR);
+	TC_ASSERT_EQ("task_restart", errno, ENOSYS);
+	TC_ASSERT_EQ("task_restart", g_callback, 0);
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_task.c]: 1. Previous logic is better. Calling ```task_restart``` is unnecessary test without any launched task.
```
	pid = task_create("tc_exit", SCHED_PRIORITY_MAX - 1, 1024, restart_task, (char * const *)NULL);
	TC_ASSERT_GT("task_create", pid, 0);

	g_callback = 0;
	ret_chk = task_restart(0);
	TC_ASSERT_EQ("task_restart", ret_chk, ERROR);
	TC_ASSERT_EQ("task_restart", errno, ENOSYS);
	TC_ASSERT_EQ("task_restart", g_callback, 0);
```

2. As I left, you need to initialize checking value, g_callback.
  As you add new testcase before testing below, if someone adds new tc before below, we can't guarantee g_callback is zero.

Comment (thapav) [reply apps/examples/testcase/le_tc/kernel/tc_task.c]: ok understood why re-initialize it,  will make these changes

File: apps/examples/testcase/le_tc/kernel/tc_task.c (hunk 11 above)
Comment (thapav) [reply apps/examples/testcase/le_tc/kernel/tc_task.c]: I checked with schedforeach(), but i am finding a discrepency 
So i will update task_create TC in a new patch.

**Comment by thapav** [comment]
modified as per review comments, please check

**Comment by seinfra** [comment]
Code Rule Check OK.

File: apps/examples/testcase/le_tc/kernel/tc_task.c (hunk 9 above)
Comment (sunghan-chang) [reply apps/examples/testcase/le_tc/kernel/tc_task.c]: check this again

File: apps/examples/testcase/le_tc/kernel/tc_task.c [hunk 13]
```diff
@@ -282,6 +294,20 @@ static void tc_task_task_restart(void)
...
 	unsigned int remain;
 	g_callback = 0;
+
+	/* Check for NULL pid parameter  */
+
+	pid = task_create("tc_task_re", SCHED_PRIORITY_MAX - 1, 1024, restart_task, (char * const *)NULL);
+	TC_ASSERT_GT("task_create", pid, 0);
+
+	g_callback = 0;
+	ret_chk = task_restart(0);
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_task.c]: To test restart functionality, below looks like correct.
1. launch A task
2. finish A task
3. restart A or test incorrect arg

But, there is no 2nd step in yours.
Even though ```restart_task``` is alive, you execute ```task_restart``` function. 
I think it is not correct.

File: apps/examples/testcase/le_tc/kernel/tc_task.c [hunk 14]
```diff
@@ -282,6 +294,20 @@ static void tc_task_task_restart(void)
...
+	TC_ASSERT_EQ("task_restart", ret_chk, ERROR);
+	TC_ASSERT_EQ("task_restart", errno, ENOSYS);
+	TC_ASSERT_EQ("task_restart", g_callback, 0);
+
+	/* Check for reinitialization of task using task_restart */
+
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_task.c]: You didn't add initialization of global variable. But you wrote "modified as you suggested". What is your opinion?

Comment (thapav) [reply apps/examples/testcase/le_tc/kernel/tc_task.c]: Hi , i misunderstood with initializing before this TC, 
I will add it after this:
        ret_chk = task_restart(0);
        TC_ASSERT_EQ("task_restart", ret_chk, ERROR);
        TC_ASSERT_EQ("task_restart", errno, ENOSYS);
        TC_ASSERT_EQ("task_restart", g_callback, 0);
        g_callback = 0;

File: apps/examples/testcase/le_tc/kernel/tc_task.c (hunk 13 above)
Comment (thapav) [reply apps/examples/testcase/le_tc/kernel/tc_task.c]: done, removed sleep(SEC_1) from restart_task entry point function

**Comment by seinfra** [comment]
Code Rule Check OK.
